### ISSUE TITLE
Income Statement, Balance Sheet and Cashflows are coming up as an empty dataframe

### ISSUE THREAD

**Issue by karthikbhamidipati** [ranaroussi/yfinance#191]
Income Statement, Balance Sheet and Cashflows are coming up as an empty data frame. 
This is because variable url in base.py is being updated to <address>/quote/<Ticker>/holders in line 283
and when fetching the financials in line 354 of base.py the url being used is <address>/quote/<Ticker>/holders/financials and it should be <address>/quote/<Ticker>/financials

**Comment by masueror** [comment]
I also have the same problem, when i ask for financials I recibe:
Empty DataFrame
Columns: [Open, High, Low, Close, Adj Close, Volume]
Index: []
How can I fix it?

**Comment by TimBstn** [comment]
Hey, I just had the same problem.
The problem is as vamshikarthik described it: Just change the following in the base.py:

 `url = "{}/{}".format(self._scrape_url, self.ticker)`
`holders = _pd.read_html(url+'\holders')`

The problem is in line 283.

Thanks vamshikarthik   for the tip.

Tim

**Comment by masueror** [comment]
Thank you very much, problem just solve

**Comment by kalinkirev** [comment]
It did not work for me I'm sorry. This applies for balance_sheet, cashflow, earnings, financials etc.

```
---------------------------------------------------------------------------
HTTPError                                 Traceback (most recent call last)
<ipython-input-8-3cf019396317> in <module>
----> 1 amzn.balance_sheet

~/anaconda3/envs/stok/lib/python3.7/site-packages/yfinance/ticker.py in balance_sheet(self)
    164     @property
    165     def balance_sheet(self):
--> 166         return self.get_balancesheet()
    167 
    168     @property

~/anaconda3/envs/stok/lib/python3.7/site-packages/yfinance/base.py in get_balancesheet(self, proxy, as_dict, freq)
    441 
    442     def get_balancesheet(self, proxy=None, as_dict=False, freq="yearly"):
--> 443         self._get_fundamentals(proxy)
    444         data = self._balancesheet[freq]
    445         if as_dict:

~/anaconda3/envs/stok/lib/python3.7/site-packages/yfinance/base.py in _get_fundamentals(self, kind, proxy)
    282         # holders
    283         url = "{}/{}/holders".format(self._scrape_url, self.ticker)
--> 284         holders = _pd.read_html(url+'\holders')
    285         self._major_holders = holders[0]
    286         self._institutional_holders = holders[1]

~/anaconda3/envs/stok/lib/python3.7/site-packages/pandas/io/html.py in read_html(io, match, flavor, header, index_col, skiprows, attrs, parse_dates, thousands, encoding, decimal, converters, na_values, keep_default_na, displayed_only)
   1103         na_values=na_values,
   1104         keep_default_na=keep_default_na,
-> 1105         displayed_only=displayed_only,
   1106     )

~/anaconda3/envs/stok/lib/python3.7/site-packages/pandas/io/html.py in _parse(flavor, io, match, attrs, encoding, displayed_only, **kwargs)
    910             break
    911     else:
--> 912         raise_with_traceback(retained)
    913 
    914     ret = []

~/anaconda3/envs/stok/lib/python3.7/site-packages/pandas/compat/__init__.py in raise_with_traceback(exc, traceback)
     45     if traceback == Ellipsis:
     46         _, _, traceback = sys.exc_info()
---> 47     raise exc.with_traceback(traceback)
     48 
     49 

HTTPError: HTTP Error 404: Not Found
```

Any other ideas?

Thanks in advance

**Comment by mariusgarbowski** [comment]
I had the same problem, which source was mentioned above. Fixing the problem as stated above results in the error of @kalinkirev . I resolved it by:

Line 283 (commenting out the url, because the raw url was already defined in line 279):
`        #url = "{}/{}/holders".format(self._scrape_url, self.ticker)
        holders = _pd.read_html(url+'/holders')`

and adding "/financials" in Line 354:

`data = utils.get_json(url+'/financials', proxy)`

Restart the kernel and it should work.

**Comment by destractor-edo** [comment]
Hi everyone, i'm new with python, and i need to use this function to download financial staments  for a university exam (i'm trying to do a short program).  the problem is that i don't know how modify (and find) the code above, can anyone help me?

**Comment by VaseSimion** [comment]
> I had the same problem, which source was mentioned above. Fixing the problem as stated above resulted in the error quoted by @kalinkirev .
> 
> I resolved it by:
> 
> Line 283 (commenting out the url, because the raw url was already defined in line 279) and adding "/holders" to the url+...:
> `#url = "{}/{}/holders".format(self._scrape_url, self.ticker)`
> `holders = _pd.read_html(url+'/holders')`
> 
> and adding "/financials" in Line 354:
> 
> `data = utils.get_json(url+'/financials', proxy)`
> 
> Restart the kernel and it should work.

Thanks, I did this and it fixed my issue

**Comment by chaitanya9186** [comment]
As described #111 it is working fine at version 0.1.48.

**Comment by anbento0490** [comment]
Hi @mariusgarbowski 

I tried to apply the fixes you suggested at line 283 and 354 and re-run the base.py code after restarting the kernel but it seems that I cannot use any of the get functions. I am not an expert Python user so I am very likely doing something naive here. Could you kindly provide a step by step guide for the full process or and example of how to use the functions after the lines have been edited? 

Thanks in advance for your help :)

**Comment by mikaelhug** [comment]
> Hey, I just had the same problem.
> The problem is as vamshikarthik described it: Just change the following in the base.py:
> 
> `url = "{}/{}".format(self._scrape_url, self.ticker)`
> `holders = _pd.read_html(url+'\holders')`
> 
> The problem is in line 283.
> 
> Thanks vamshikarthik for the tip.
> 
> Tim

Worked for me as well, thanks!

**Comment by FTorres09** [comment]
Hi guys, I'm rather new to Python which is why I don't quite get the methodology of your edits.
I installed yfinance for Anaconda which is the place I need it for. Unfortunately, when I install it at Anaconda, the financials part also doesn't work on my Mac. Which is the reason why I'd like to ask y'all what to do in order to solve that problem. 
I haven't downloaded yfinance on my computer, only got it from the base (root) Anaconda environment.

**Comment by aniketnaikdesai** [comment]
Copied from @anbento0490 

I am going to explain in depth the method I used to fix it for less advanced git_hub/python users:

    First of all, fork yfinance repository in your account. In this way you will be allowed to modify the code as you wish,
    In your account, in the forked repository, open base.py file and fix lines 283 and 284:
    283 url = "{}/{}".format(self._scrape_url, self.ticker) 284 holders = _pd.read_html(url+'/holders')
    Commit the changes
    If you had yfinance installed, uninstall the original version
    Finally, pip install the package directly from your gitub account with the following command:

pip install git+https://github.com/username/yfinance.git

After you do that, the methods that were failing will work as expected!

**Comment by FTorres09** [comment]
Thank you so much @aniketnaikdesai! For the terminal it worked perfectly. Is there also a solution available for Anaconda (Navigator)? It's not impossible to install it there with 
`conda install git+https://github.com/username/yfinance.git`
I would try to work out some yfinance calcs there.
Thanks a great lot in advance for your help!

**Comment by rodrigobercini** [comment]
> Copied from @anbento0490
> 
> I am going to explain in depth the method I used to fix it for less advanced git_hub/python users:
> 
> ```
> First of all, fork yfinance repository in your account. In this way you will be allowed to modify the code as you wish,
> In your account, in the forked repository, open base.py file and fix lines 283 and 284:
> 283 url = "{}/{}".format(self._scrape_url, self.ticker) 284 holders = _pd.read_html(url+'/holders')
> Commit the changes
> If you had yfinance installed, uninstall the original version
> Finally, pip install the package directly from your gitub account with the following command:
> ```
> 
> pip install git+https://github.com/username/yfinance.git
> 
> After you do that, the methods that were failing will work as expected!

Thank you, this fix worked for me.

There was only an additional step: I had to delete the file "Icon" from the repo, it was causing an installation issue via pip. You can use my own fork if you want:

pip install git+https://github.com/rodrigobercini/yfinance.git

**Comment by aniketnaikdesai** [comment]
I don't think you can conda install from github.

**Comment by phsamuel** [comment]
Some symbols work but many symbols still don't work after the fix @aniketnaikdesai. 
For example, I can access the symbol AB in yahoo finance. But I can't access the balance sheet with yfinance. It returns an index out of range error. 

![image](https://user-images.githubusercontent.com/5240193/79305922-dbce4780-7eb9-11ea-8501-43f0332798c7.png)

**Comment by aniketnaikdesai** [comment]
I am able to pull the balance sheet information for AB. I have my code patched for many changes.

You can also fix your code.

The index error is due to missing institutional holders information.
Look for the fix in issue #229 or #159
but there seems to be a _get_fundamentals error also. Look at issue #174

**Comment by phsamuel** [comment]
@aniketnaikdesai Thanks much. It is very helpful!

**Comment by TomData** [comment]
> 
> 
> As described #111 it is working fine at version 0.1.48.

I am using version 0.1.54 and still have the issues described above. Am currently trying the fixes from above

**Comment by martinemilander** [comment]
`pip install git+https://github.com/username/yfinance.git` isn't working for me, for some reason. I get the following error:

ERROR: Command errored out with exit status 1: git clone -q https://github.com/martinemilander/yfinance.git /private/var/folders/3c/nkrc6kr94p54njs4j7333wq80000gn/T/pip-req-build-3yued3xj Check the logs for full command output.

It then prompts me to install command line tools for developers (something to do with Xcode) but then the installation fails.

**Comment by waterproofpatch** [comment]
I recently started using this library and ran into the issue described by @karthikbhamidipati. The change I made in base.py was to change line 376 to remove the superflous '/holders' URI component, such that the resultant URL suffixes with /<ticker>/financials, rather than /<ticker>/holders/financials.

I can now use .financials, .balance_sheet and .earnings for the Ticker object, for at least the one stock Ticker I've tested with.

Attached is the diff for my change.

[patch.txt](https://github.com/ranaroussi/yfinance/files/5897335/patch.txt)


Hope this helps someone else.
